domain/src/main/java/org/oppia/domain/oppialogger/LogReportStorageHelper.kt

```diff
@@ -0,0 +1,145 @@
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.launch
+import org.oppia.app.model.EventLog
+import org.oppia.app.model.ExceptionLog
+import org.oppia.app.model.OppiaEventLogs
+import org.oppia.app.model.OppiaExceptionLogs
+import org.oppia.data.persistence.PersistentCacheStore
+import org.oppia.util.data.AsyncResult
+import org.oppia.util.data.DataProviders
+import org.oppia.util.logging.Logger
+import org.oppia.util.threading.BackgroundDispatcher
+import javax.inject.Inject
+import javax.inject.Singleton
+
+@Singleton
+class LogReportStorageHelper @Inject constructor(
+  private val cacheStoreFactory: PersistentCacheStore.Factory,
+  private val dataProviders: DataProviders,
+  private val logger: Logger,
+  @BackgroundDispatcher private val backgroundCoroutineDispatcher: CoroutineDispatcher,
+  @EventLogStorageCacheSize private val eventLogStorageCacheSize: Int,
+  @ExceptionLogStorageCacheSize private val exceptionLogStorageCacheSize: Int
+) {
+  private enum class LogReportingCase {
+    EVENT_LOG,
+    EXCEPTION_LOG
+  }
+  private val eventLogStore =
+    cacheStoreFactory.create("event_logs", OppiaEventLogs.getDefaultInstance())
+  private val exceptionLogStore =
+    cacheStoreFactory.create("exception_logs", OppiaExceptionLogs.getDefaultInstance())
+  private val coroutineScope = CoroutineScope(backgroundCoroutineDispatcher)
+
+  fun addEventLog(eventLog: EventLog) {
+    coroutineScope.launch {
+      checkStoreCacheStatus(
+        eventLogStore.readDataAsync().await().serializedSize.toByte(),
+        LogReportingCase.EVENT_LOG,
+        eventLogStorageCacheSize
+      )
+    }
+    eventLogStore.storeDataAsync(updateInMemoryCache = true) {
+      it.toBuilder().addEventLog(eventLog).build()
+    }.invokeOnCompletion {
+      it?.let {
+        logger.e(
+          "DOMAIN",
+          "Failed to store event log",
+          it
+        )
+      }
+    }
+  }
+
+  fun addCrashLog(exceptionLog: ExceptionLog) {
+    coroutineScope.launch {
+      checkStoreCacheStatus(
+        exceptionLogStore.readDataAsync().await().serializedSize.toByte(),
+        LogReportingCase.EXCEPTION_LOG,
+        exceptionLogStorageCacheSize
+      )
+    }
+    exceptionLogStore.storeDataAsync(updateInMemoryCache = true) {
+      it.toBuilder().addExceptionLog(exceptionLog).build()
+    }.invokeOnCompletion {
+      it?.let {
+        logger.e(
+          "DOMAIN",
+          "Failed to store crash log",
+          it
+        )
+      }
+    }
+  }
+
+  private suspend fun checkStoreCacheStatus(
+    storeSize: Byte,
+    logReportCase: LogReportingCase,
+    cacheStorageLimit: Int
+  ) {
+    if (storeSize / 1048576 > cacheStorageLimit) {
+      when (logReportCase) {
+        LogReportingCase.EVENT_LOG -> removeEvent(getLeastRecentEvent())
+        LogReportingCase.EXCEPTION_LOG -> removeException(getLeastRecentException())
+      }
+      checkStoreCacheStatus(storeSize, logReportCase, cacheStorageLimit)
```

Comment:
Why exactly are we doing recursive method calls here? I am not sure if that's a good idea. Instead we can pass in how much of data in bytes needs to be removed (N = currentSize - cacheSize) and take the leastRecentEventsUptoSizeN and remove it in one shot. 
Alternatively, you can keep the cache value as number of records instead of bytes, and then you can remove leastRecentNEvents if the size of the cache becomes greater than the total number of entries allowed by N

Reply:
Yes, the records method seems like a better approach. Modified, PTAL.